### ISSUE TITLE
Support dry-running in `TestExternalities`

### ISSUE THREAD

**Issue by pmikolajczyk41** [paritytech/substrate#14825]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Motivation

In some cases where we are working with `TestExternalities` it is required to perform some simulation, i.e. some action that eventually won't influence storage. For examples see:
 - [`try-runtime::fast-forward`](https://github.com/paritytech/substrate/blob/6b07b97358abb4ae4e132c11579c3ed52cdd648c/utils/frame/try-runtime/cli/src/commands/fast_forward.rs#L227)
 - [estimating gas limit for drink! backend in ink! E2E tests](https://github.com/Cardinal-Cryptography/drink/pull/39)

The problem is that there is no reliable API for doing so. Here are some ideas and their blockers:
 - using `into_raw_snapshot` and `from_raw_snapshot`: consumes `self` and thus doesn't allow for backup-simulation-restore scenario; also, drops extensions
 - saving backend with `as_backend()` and overwriting a field in the `TestExternalities` object: requires manual intervention into caches, offchain db, overlay; not so sure whether `commit_all()` before restoring is sufficient
 - saving `Storage` object and recreating `TestExternalities` with it: there's no way of obtaining raw `Storage` object; also, drops extensions

### Request

Expose API enabling 'backup-simulate-restore' mechanism for `TestExternalities`.

### Solution

If I'm not mistaken, it would be enough to expose method for exporting `Storage` and `Extensions` data and then creating new externalities out of them.

### Are you willing to help with this request?

Yes!

**Comment by ggwpez** [comment]
Can be done as a macro like here:  
https://github.com/paritytech/substrate/blob/oty-safe-mode/frame/safe-mode/src/tests.rs#L31-L50

Usage:  
```rust
Storage::<T>::set(4);
hypothetically!({
    Storage::<T>::set(5);
    assert_eq!(5, Storage::<T>::get());
});
assert_eq!(4, Storage::<T>::get());
```

Then it works with all externalities by just using the storage overlay.

**Comment by ggwpez** [comment]
> we would have to somehow capture the returning value from the simulation process

It is supported like this:  
```rust
let ret = hypothetically!({
	123
});
assert_eq!(123, ret);
```

We can pull it out of this MR if you want it https://github.com/paritytech/substrate/pull/12092  
But i would wait until Monday after the monorepo migration.

**Comment by pmikolajczyk41** [comment]
sure, let's do it that way; but I'm wondering whether such macro shouldn't be moved from a pallet to e.g. sp-state-machine...

**Comment by ggwpez** [comment]
> sure, let's do it that way; but I'm wondering whether such macro shouldn't be moved from a pallet to e.g. sp-state-machine...

Yes it should. I was just lazy in that MR and did not want to justify changes to FRAME. So i put it into the test module.
